### PR TITLE
attachedCluster: add controller and enable to be registered with Fleet.

### PR DESCRIPTION
**Pull request by Xieql**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

- add attachedcluster controller functionality.
- enable attachedclusters to be registered with Fleet.
- check if the cluster is already joined in fleet.
- removed the "accepted" field in "attachedcluster". because the fleet label can indicates it.
- removed the fleet label when the the cluster is unjoined.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

### CODE REVIEW

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
We donot need a controller for attached cluster

File: pkg/fleet-manager/clusters.go [hunk 1]
```diff
@@ -36,20 +40,47 @@ import (
...
 		clusterKey := types.NamespacedName{Name: cluster.Name, Namespace: fleet.Namespace}
-		var cluster clusterv1alpha1.Cluster
-		err := f.Get(ctx, clusterKey, &cluster)
+		var currentCLuster ClusterInterface
+		var err error
+		if cluster.Kind == ClusterKind {
+			var cluster clusterv1alpha1.Cluster
+			err = f.Get(ctx, clusterKey, &cluster)
+			currentCLuster = &cluster
+		} else if cluster.Kind == AttachedClusterKind {
```
Comment (hzxuzhonghu) [inline pkg/fleet-manager/clusters.go:72]: nit: We can also support custom cluster, not a must now

File: pkg/apis/cluster/v1alpha1/cluster_types.go [hunk 2]
```diff
@@ -263,3 +264,23 @@ type ClusterList struct {
...
+	return c
+}
+
+func (c *Cluster) GetSecretName() string {
+	return c.Spec.Credential.SecretRef
+}
+
+// ClusterKubeconfigDataName is the key used to store a Kubeconfig in the secret's data field.
+// This is derived from cluster api
+const ClusterKubeconfigDataName = "value"
```
Comment (hzxuzhonghu) [inline pkg/apis/cluster/v1alpha1/cluster_types.go:282]: good

File: pkg/controllers/attachedcluster_controller.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	attachedClusterConfig, err := clientcmd.RESTConfigFromKubeConfig(secret.Data[attachedCluster.GetSecretKey()])
+	if err != nil {
+		log.Error(err, "build restconfig for controlplane failed")
+		return ctrl.Result{}, fmt.Errorf("build restconfig for controlplane failed %v", err)
+	}
+	controlPlaneKubeClient := kubeclient.NewForConfigOrDie(attachedClusterConfig)
+
+	// The default cluster ID for Karmada is the UID of the NamespaceSystem in the cluster.
```
Comment (hzxuzhonghu) [inline pkg/controllers/attachedcluster_controller.go:105]: remove, this is not relevant to karmada

**Comment by Xieql** [comment]
> We donot need a controller for attached cluster

Yes, it can work without a controller for the attached cluster. 

However, adding this controller enables us to achieve separation of concerns and preview checks in the attached cluster controller. This includes checking the accessibility of the secret for the attached cluster, validating if the secret can be transformed into kubeconfig, and verifying if the cluster recorded in kubeconfig is accessible. By doing so, we can avoid reading error information in the fleet manager.

Additionally, remote cluster accessibility validation can be a time-consuming action in some situations. Asynchronous validation in attahced cluster controllers can avoid invalid cluster and  enhance the system's efficiency and responsiveness.

File: pkg/fleet-manager/clusters.go (hunk 1 above)
Comment (Xieql) [reply pkg/fleet-manager/clusters.go:72]: will add this in a subsequent pr

**Review by hzxuzhonghu — APPROVED** [review]
/lgtm

**Pull request merged**